pymc3/diagnostics.py

```diff
@@ -231,27 +231,21 @@ def get_neff(x, Vhat):
         # Chain samples are second to last dim (-2)
         num_samples = x.shape[-2]
 
-        negative_autocorr = False
-
-        rho = np.ones(num_samples)
-        t = 1
-
-        # Iterate until the sum of consecutive estimates of autocorrelation is
-        # negative
-        while not negative_autocorr and (t < num_samples):
-
-            variogram = np.mean((x[t:, :] - x[:-t, :])**2)
-            rho[t] = 1. - variogram / (2. * Vhat)
-
-            negative_autocorr = sum(rho[t - 1:t + 1]) < 0
+        # Calculate within-chain variance
+        W = np.mean(np.var(x, axis=-2, ddof=1), axis=-1)
 
-            t += 1
+        rho = np.ones(2 * num_chains + 1)
 
-        if t % 2:
-            t -= 1
+        # Iterate over different lags of autocorrelation
+        for t in range(1, 2 * num_chains + 2):
+            auto_corr = []
+            for m in range(num_chains):
+                auto_corr.append(autocorr(x[:, m], t))
+            rho[t - 1] = 1. - (W - np.mean(auto_corr)) / Vhat
 
-        neff = num_chains * num_samples / (1. + 2 * rho[1:t-1].sum())
-        return min(num_chains * num_samples, np.floor(neff))
+        tHat = 1. + 2 * rho.sum()
+        neff = num_chains * num_samples / tHat
+        return abs(neff)
 
     def generate_neff(trace_values):
         x = np.array(trace_values)
```

Comment:
np.array() function gives different outputs for trace objects and  lists of arrays. This is probably the source of the problem. Will fix it.